src/jtag/drivers/remote_bitbang.c

```diff
@@ -90,20 +92,37 @@ static int remote_bitbang_fill_buf(void)
 	return ERROR_OK;
 }
 
-static int remote_bitbang_putc(int c)
+static int remote_bitbang_flush(void)
```

Comment:
Done. (I removed the assert because it always fired. I did not take the time to investigate why.)

Reply:
I am sorry, but the failing assert should not be dropped just like that. It means there is still a hidden bug in this merge request. 

I took a moment to look at it. The catch is that OpenOCD can call the `.reset` callback of `struct adapter_driver` directly, without the subsequent call to `jtag_execute_queue()`. (See: https://github.com/riscv/riscv-openocd/blob/riscv/src/jtag/core.c#L814) As a result, the reset command ('r' character) is buffered and not sent out, remote_bitbang_send_buf_used==1, which fires the assert when the JTAG queue is next executed. 

Flushing the queue on reset should fix it:
```
diff --git a/src/jtag/drivers/remote_bitbang.c b/src/jtag/drivers/remote_bitbang.c
index 214fa43fc..43452766b 100644
--- a/src/jtag/drivers/remote_bitbang.c
+++ b/src/jtag/drivers/remote_bitbang.c
@@ -203,7 +203,9 @@ static int remote_bitbang_write(int tck, int tms, int tdi)
 static int remote_bitbang_reset(int trst, int srst)
 {
        char c = 'r' + ((trst ? 0x2 : 0x0) | (srst ? 0x1 : 0x0));
-       return remote_bitbang_queue(c, false);
+       /* Always flush the send buffer on reset, because the reset call
+          need not be followed by jtag_execute_queue(). */
+       return remote_bitbang_queue(c, true);
 }
```

Hope this helped.

Reply:
Fixed. Thank you for tracking that down.